### PR TITLE
fix(providers): keep Responses content vocabulary on explicit prompt-cache breakpoints

### LINKED ISSUE #861: Issue: explicit prompt-cache breakpoints send the Chat content vocabulary on /v1/responses, so OpenAI rejects them with 400
---

**Describe the bug**

When the prompt-cache planner attaches an explicit cache breakpoint to a `/v1/responses` request, it builds the content part as a `core.ContentPart`. That is the **Chat** content type: `ContentPart.MarshalJSON` (`internal/core/chat_content.go:83`) hardcodes the emitted type, so `Type: "input_text"` is serialized as `"type":"text"`.

```go
case "text", "input_text":
    return marshalWithUnknownJSONFields(struct {
        Type string `json:"type"`
        Text string `json:"text"`
    }{
        Type: "text",   // <- "input_text" is normalized away here
        Text: p.Text,
    }, p.ExtraFields)
```

The Responses API rejects that vocabulary, so the request fails with a 400 before reaching the model. This is request-schema validation, not model behavior, so it fires for every model matching `supportsExplicitOpenAICache` (`strings.Contains(model, "gpt-5.6")`) whose prefix crosses the 1024-token OpenAI minimum in `providerCacheMinimum`.

Two branches of `markOpenAIResponsesBreakpoint` synthesize the wrong type:

- `internal/providers/cache_planner.go:344` — `case string` builds `[]core.ContentPart{{Type: "input_text", Text: content}}`
- `internal/providers/cache_planner.go:351` — `case []core.ContentPart` attaches the breakpoint through `ExtraFields` on a `ContentPart`

The `case []any` and `case []map[string]any` branches are fine. Real `/v1/responses` ingress decodes message content into generic values (`ResponsesInputElement.UnmarshalJSON`, `internal/core/responses_json.go`), so those paths preserve whatever vocabulary the client sent — only the two `ContentPart` branches introduce it.

Root cause in one line: `core.ContentPart` is the Chat-direction type and has no business in a Responses request. The repository already has the right counterpart, `core.ResponsesContentItem` (`internal/core/responses.go:215`), which serializes `Type` verbatim — but it has no `ExtraFields`, so there is nowhere to hang `prompt_cache_breakpoint`, which is presumably why the planner reached for `ContentPart`.

**GoModel version**

`upstream/main` @ `a8fe362e` (`git describe`: `v0.1.85-4-ga8fe362e`)

**Steps To Reproduce (optionally)**

The API rejection reproduces directly against OpenAI, without the gateway:

1. Send a Responses request whose input content part uses the Chat type:

```bash
curl -sS -w '\nHTTP %{http_code}\n' https://api.openai.com/v1/responses \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -H 'Content-Type: application/json' \
  -d '{"model":"<a gpt-5.6 model>","max_output_tokens":16,
       "input":[{"role":"user","content":[{"type":"text","text":"ping"}]}]}'
```

2. Observe HTTP 400 (log below).
3. Repeat with `"type":"input_text"` — HTTP 200, `status: "completed"`.

Through the gateway, the same payload is produced whenever the planner fires: a `gpt-5.6` model, a prefix above 1024 tokens, and at least one further input item after it.

The existing unit test does not catch this. `TestCachePlannerResponsesShapesAndCallerOwnership` covers all three content shapes but only asserts that `prompt_cache_breakpoint` appears somewhere in the marshalled request, never the emitted part type. Adding these assertions inside the existing loop fails on the `string` and `typed parts` shapes:

```go
// The Responses API rejects the Chat content vocabulary, so a breakpoint
// must never downgrade a part to {"type":"text"}.
if bytes.Contains(plannedJSON, []byte(`"type":"text"`)) {
    t.Fatalf("Responses plan emitted Chat content vocabulary: %s", plannedJSON)
}
if !bytes.Contains(plannedJSON, []byte(`"type":"input_text"`)) {
    t.Fatalf("Responses plan lost the input_text vocabulary: %s", plannedJSON)
}
```

**Screenshots/Logs (optionally)**

Step 2, HTTP 400:

```json
{
  "error": {
    "message": "Invalid value: 'text'. Supported values are: 'input_text', 'input_image', 'input_audio', 'output_text', 'refusal', 'input_file', 'computer_screenshot', 'summary_text', and 'encrypted_content'.",
    "type": "invalid_request_error",
    "param": "input[0].content[0].type",
    "code": "invalid_value"
  }
}
```

Step 3 control, HTTP 200:

```json
{
  "id": "resp_…",
  "object": "response",
  "status": "completed",
  "model": "<a gpt-5.6 model>",
  "output": [{"type":"message","role":"assistant",
              "content":[{"type":"output_text","text":"pong"}]}]
}
```

The accepted set mixes input and output kinds because the Responses API carries request input and response output in one item stream — prior output items can be fed straight back as input. The type therefore has to encode direction, and a bare `"text"` is not resolvable. Chat Completions never needed this, because position (`messages[]` vs `choices[]`) disambiguated it — which is exactly the assumption baked into `ContentPart`.

**Browser (if applicable)**

n/a — reproduced with curl and `go test`.

---

**Suggested fix**

1. **Add `ExtraFields UnknownJSONFields` to `ResponsesContentItem`** and have the planner build Responses items with it. This fixes the class rather than one call site, and gives the Responses path the unknown-field round-tripping the Chat path already has (consistent with the forwarding rules recorded in the ADR from #857). It also unblocks anything that needs to annotate Responses content — relevant to #723, where Chat↔Responses translation makes the `ContentPart` / `ResponsesContentItem` asymmetry structural rather than incidental.
2. **Build raw `map[string]any` blocks in the planner**, mirroring what the `[]any` branch already manipulates. Smaller and local to `cache_planner.go`, but leaves `core` unable to express an annotated Responses content part.

Other sites placing `[]core.ContentPart` into `ResponsesInputElement.Content` are worth auditing too. `internal/guardrails/responses_message_apply.go:442` constructs `[]core.ContentPart{{Type: "input_text", Text: rewrittenText}}` for Responses input and looks like the same class of defect; I have not verified whether that value reaches provider serialization unconverted.

### PR DESCRIPTION
**Pull request by SantiagoDePolonia**
## Description

Fixes #861.

When the prompt-cache planner attached an explicit breakpoint to a `/v1/responses` request, it built the marked part as a `core.ContentPart`. That is the Chat content type, and its `MarshalJSON` rewrites `input_text` to `text`. The Responses API rejects that vocabulary, so every planned request for a `gpt-5.6` model with a prefix over the 1024-token OpenAI minimum failed with a 400 before reaching the model.

Changes:

- `core.ResponsesBlocksFromContentParts` converts typed parts into generic blocks with their `type` kept verbatim. It is the one place that knows typed parts must not be serialized directly on the Responses path.
- The planner marks generic map blocks on every content shape. The three per-shape marking loops are folded into one helper, and typed parts go through the new conversion.
- The guardrails rewrite for typed Responses content had the same defect: it prepended a typed `input_text` part that would serialize as `text`. That branch now converts to blocks and delegates to the existing map-based rewrite, which removes the duplicated typed implementation. This path is only reachable with in-process typed content, so it was latent rather than user-visible.

User-visible impact: explicit prompt caching on `/v1/responses` for OpenAI `gpt-5.6` models works again. No other provider or path changes.

Tests: the Responses shape test asserts the emitted part type and exact breakpoint block, and fails on `main` for the string and typed-parts shapes. New tests cover typed parts carrying extra fields and an `input_image` part, the already-marked and unencodable edge cases, the core helper, and the guardrails typed rewrite.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved OpenAI Responses API content handling while preserving Responses-specific content types and additional fields.
  * Prompt-cache breakpoints are now applied to the last eligible content block, including typed text, image, and file content.
  * Unsupported or unencodable content is handled safely without producing invalid requests.
  * Existing cache breakpoints remain unchanged.

### CODE REVIEW

File: internal/core/responses_json.go [hunk 1]
```diff
@@ -363,6 +363,29 @@ func (c ResponsesContentItem) MarshalJSON() ([]byte, error) {
...
+// instead of serializing them directly.
+func ResponsesBlocksFromContentParts(parts []ContentPart) []any {
+	blocks := make([]any, 0, len(parts))
+	for _, part := range parts {
+		encoded, err := json.Marshal(part)
+		if err != nil {
+			return nil
+		}
+		var block map[string]any
+		if err := json.Unmarshal(encoded, &block); err != nil {
```
[the remark this answers is not in the document]
Comment (SantiagoDePolonia) [reply internal/core/responses_json.go]: Addressed in 251cea2f: the block decode now uses UseNumber, with a regression test for an integer above 2^53.
